### PR TITLE
issue: 842842 Move udp rx hw timestamp conversion from cq_mgr into so…

### PR DESCRIPTION
**Pull request by liranoz12**
…ckinfo_udp

Currently, The time conversion for each incoming packet is made in the cq_mgr.
i.e, the conversion is made no only for UDP sockets who didn't enable the option
but also for TCP sockets.
Each conversion includes some mathematical calculations and cache misses
which cause performance degradation.
To improve performance, we should make this conversion only for UDP
sockets which enabled the option.

Signed-off-by: Liran Oz <lirano@mellanox.com>

### CODE REVIEW

**Comment by liranoz12** [comment]
This is a new PR based on - https://github.com/Mellanox/libvma/pull/162
@rosenbaumalex @ilansmith @OphirMunk  Please review

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/672/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/664/ for details (Mellanox internal link).

File: src/vma/sock/sockinfo_udp.h [hunk 1]
```diff
@@ -46,6 +46,7 @@
 #include "vma/util/sys_vars.h"
 #include "vma/proto/mem_buf_desc.h"
 #include "vma/proto/dst_entry_udp.h"
+#include "vma/dev/ring_simple.h"
```
Comment (ilansmith) [inline src/vma/sock/sockinfo_udp.h]: Why is this required here?

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/675/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/679/ for details (Mellanox internal link).

File: src/vma/dev/cq_mgr.h [hunk 2]
```diff
@@ -168,9 +168,7 @@ class cq_mgr
 
 	void 	modify_cq_moderation(uint32_t period, uint32_t count);
 
-	inline void convert_hw_time_to_system_time(uint64_t packet_hw_time, struct timespec* packet_systime) {
-		m_p_ib_ctx_handler->convert_hw_time_to_system_time(packet_hw_time, packet_systime);
-	}
+	inline void convert_hw_time_to_system_time(uint64_t hwtime, struct timespec* systime) { m_p_ib_ctx_handler->convert_hw_time_to_system_time(hwtime, systime); }
```
Comment (OphirMunk) [inline src/vma/dev/cq_mgr.h]: ring already knows ib_cxt_handler. why going through the ring?

File: src/vma/dev/cq_mgr.cpp [hunk 3]
```diff
@@ -600,7 +600,7 @@ mem_buf_desc_t* cq_mgr::process_cq_element_rx(vma_ibv_wc* p_wce)
 		//this is not a deadcode if timestamping is defined in verbs API
 		// coverity[dead_error_condition]
 		if (vma_wc_flags(*p_wce) & VMA_IBV_WC_WITH_TIMESTAMP) {
-			m_p_ib_ctx_handler->convert_hw_time_to_system_time(vma_wc_timestamp(*p_wce) ,&p_mem_buf_desc->path.rx.hw_timestamp);
+			p_mem_buf_desc->path.rx.hw_raw_timestamp = vma_wc_timestamp(*p_wce);
```
Comment (OphirMunk) [inline src/vma/dev/cq_mgr.cpp]: Use in place: use hw_timestamp and remove hw_raw_timestamp

File: src/vma/dev/cq_mgr.cpp [hunk 4]
```diff
@@ -667,6 +667,7 @@ void cq_mgr::reclaim_recv_buffer_helper(mem_buf_desc_t* buff)
 				temp->path.rx.sw_timestamp.tv_sec = 0;
 				temp->path.rx.hw_timestamp.tv_nsec = 0;
 				temp->path.rx.hw_timestamp.tv_sec = 0;
+				temp->path.rx.hw_raw_timestamp = 0;
```
Comment (OphirMunk) [inline src/vma/dev/cq_mgr.cpp]: can remove it?

File: src/vma/dev/cq_mgr.h [hunk 5]
```diff
@@ -168,6 +168,10 @@ class cq_mgr
 
 	void 	modify_cq_moderation(uint32_t period, uint32_t count);
 
+	inline void convert_hw_time_to_system_time(uint64_t packet_hw_time, struct timespec* packet_systime) {
```
Comment (OphirMunk) [inline src/vma/dev/cq_mgr.h]: Try calling ib_cxt_handler directly from ring

File: src/vma/sock/sockinfo_udp.cpp [hunk 6]
```diff
@@ -1839,6 +1839,15 @@ bool sockinfo_udp::rx_input_cb(mem_buf_desc_t* p_desc, void* pv_fd_ready_array)
 		!p_desc->path.rx.sw_timestamp.tv_sec) {
 		clock_gettime(CLOCK_REALTIME, &(p_desc->path.rx.sw_timestamp));
 	}
+
+	// convert hw timestamp to system time
+	if (m_n_tsing_flags & SOF_TIMESTAMPING_RAW_HARDWARE) {
+		ring_simple* base_ring = (ring_simple*) p_desc->p_desc_owner;
+		if (base_ring) {
+			base_ring->convert_hw_time_to_system_time(p_desc->path.rx.hw_raw_timestamp, &p_desc->path.rx.hw_timestamp);
```
Comment (OphirMunk) [inline src/vma/sock/sockinfo_udp.cpp]:
```
base_ring->convert_hw_time_to_system_time(p_desc->path.rx.hw_timestamp, &p_desc->path.rx.hw_timestamp);
```

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/682/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
bot:retest

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/687/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/accl-libvma-pr/677/ for details (Mellanox internal link).

**Comment by OphirMunk** [comment]
Merged #257

**Comment by mellanox-github** [comment]
Test PASSed.
See http://bgate.mellanox.com/jenkins/job/accl-libvma-pr/688/ for details (Mellanox internal link).
